pos-voucher-modal/pos-voucher-modal.page.ts

```diff
@@ -1,45 +1,195 @@
+    public pageProvider: PR_ProgramProvider,
+    public programPartnerProvider: PR_ProgramPartnerProvider,
+    public programItemProvider: PR_ProgramItemProvider,
+    public deductionProvider:SALE_OrderDeductionProvider,
     public env: EnvService,
     public modalController: ModalController,
     public loadingController: LoadingController,
   ) 
   {
     super();
   }
-  loadData(event) {
-    this.VoucherDiscountAmount = this.item.OriginalDiscount1;
-  }
-  changeVoucher(){
-    // this.voucherProvider.read({Code:this.VoucherCode}).then(result=>{
-    //   this.Program = result['data'][0];         
-    // }).catch(err=>{});
-    
-  }
-  applyVoucher(apply = false){
-    
-    // if(this.Program.MinOrderValue > this.item.OriginalTotalBeforeDiscount){
-    //   this.env.showMessage("đơn hàng của bạn chưa đủ điều kiện áp dụng mã voucher này","warning");
-    //   return;
-    // }
-    // this.item.OriginalDiscount1 = this.Program.Value;
-    // return this.modalController.dismiss(this.item, (apply ? 'confirm' : 'cancel'));
+  loadData(event?: any): void {
+    let date =  new Date();
+    date.setHours(0,0,0,0);
+    Object.assign(this.query, {
+      IsPublic: true,
+      IsDeleted:false,
+      BetweenDate: date,
+      Type:"Voucher",
+      CanUse: true,
+      Status: 'Approved',
+    });
+    super.loadData();
+  }
+  
+  loadedData(event?: any, ignoredFromGroup?: boolean): void {
+    super.loadedData();
+    this.loadProgram();
+  }
+  loadProgram(){
+    this.items.forEach(i=>{
+      i.Used = false;
+      let find = this.item.Deductions.find(p=>p.IDProgram== i.Id);
+      if(find){
+        i.Used = true;
+      }
+      if(i.IsByPercent == true){
+        i.Value = i.Value * this.item.OriginalTotalBeforeDiscount / 100;
+        if(i.Value > i.MaxValue){
+          i.Value = i.MaxValue;
+        }
+      }
+    })
+  }
+  changeCode(){
+    if(this.Code != ""){
+      let date =  new Date();
+      date.setHours(0,0,0,0);
+      
+      let query = {
+        Code_eq:this.Code,
+        BetweenDate: date,
+        Type:"Voucher",
+        CanUse: true,
+        Status: 'Approved',
+      }
+      this.pageProvider.read(query).then(result=>{
+        if(result['count']>0)
+        {
+            this.Voucher = result['data'][0];
+            this.Voucher.Used = false;
+            let find = this.item.Deductions.find(p=>p.IDProgram== this.Voucher.Id);
+            if(find){
+              this.Voucher.Used = true;
+            }
+            if(this.Voucher.IsByPercent == true){
+              this.Voucher.Value = this.Voucher.Value * this.item.OriginalTotalBeforeDiscount / 100;
+              if(this.Voucher.Value > this.Voucher.MaxValue){
+                this.Voucher.Value = this.Voucher.MaxValue;
+              }
+            }       
+        }
+        else{
+          this.env.showMessage("Mã Voucher không hợp lệ","danger");
+        }
+      }).catch(err=>{});
+    }
+  }
+  async applyVoucher(line){
```

Comment:
Không apply như vầy được
Xử lý bên backend
vd: POST lên api VoucherApply 2 params IDSaleOrder, IDVoucher
=> Backend phải tự kiểm tra + tính toán + update cho SO => response refresh lại